packages/dds/tree/src/core/undo/undoRedoManager.ts

```diff
@@ -0,0 +1,122 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { ChangeFamily, ChangeFamilyEditor } from "../change-family";
+import { GraphCommit, GraphCommitType, RevisionTag, tagChange } from "../rebase";
+import { ReadonlyRepairDataStore, RepairDataStore } from "../repair";
+import { Delta } from "../tree";
+
+/**
+ * Manages the undo commit tree and repair data associated with undoable commits.
+ */
+export class UndoRedoManager<TChange, TEditor extends ChangeFamilyEditor> {
+	private pendingRepairData?: RepairDataStore;
+
+	/**
+	 * @param repairDataStoryFactory - Factory function for creating {@link RepairDataStore}s to create and store repair
+	 * data for {@link UndoableCommit}s.
+	 * @param changeFamily - {@link ChangeFamily} used for inverting changes.
+	 * @param applyChange - Callback to apply undos as local changes. This should call {@link UndoRedoManager.trackCommit}
+	 * with the created commit.
+	 * @param headUndoCommit - Optional commit to set as the initial undoable commit.
+	 */
+	public constructor(
+		private readonly repairDataStoryFactory: () => RepairDataStore,
+		private readonly changeFamily: ChangeFamily<TEditor, TChange>,
+		protected readonly applyChange: (
+			change: TChange,
+			revision?: RevisionTag,
+			type?: GraphCommitType,
+		) => void,
```

Comment:
Rather than take this function in, I wonder if the code might be neater if we just have the `undo` method return `TChange | undefined`.
The caller of `undo` can then call `applyChange` with it. It may make unit testing a little easier too.

Reply:
This is how I had it originally but there were some reasons I went with this method instead that have since gone away so yeah, I agree we should just return the change :)